core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -154,17 +162,179 @@ class ControllerApis(val requestChannel: RequestChannel,
+    val duplicatedTopicNames = new util.HashSet[String]
+    val topicNamesToResolve = new util.HashSet[String]
+    val topicIdsToResolve = new util.HashSet[Uuid]
+    val duplicatedTopicIds = new util.HashSet[Uuid]
+
+    def appendResponse(name: String, id: Uuid, error: ApiError): Unit = {
+      responses.add(new DeletableTopicResult().
+        setName(name).
+        setTopicId(id).
+        setErrorCode(error.error.code).
+        setErrorMessage(error.message))
+    }
+
+    def maybeAppendToTopicNamesToResolve(name: String): Unit = {
+      if (duplicatedTopicNames.contains(name) || !topicNamesToResolve.add(name)) {
+        appendResponse(name, ZERO_UUID, new ApiError(INVALID_REQUEST, "Duplicate topic name."))
+        topicNamesToResolve.remove(name)
+        duplicatedTopicNames.add(name)
+      }
+    }
+
+    def maybeAppendToIdsToResolve(id: Uuid): Unit = {
+      if (duplicatedTopicIds.contains(id) || !topicIdsToResolve.add(id)) {
+        appendResponse(null, id, new ApiError(INVALID_REQUEST, "Duplicate topic ID."))
+        topicIdsToResolve.remove(id)
+        duplicatedTopicIds.add(id)
+      }
+    }
+
+    request.topicNames.forEach(maybeAppendToTopicNamesToResolve)
+
+    request.topics.forEach {
+      topic => if (topic.name == null) {
+        if (topic.topicId.equals(ZERO_UUID)) {
+          appendResponse(null, ZERO_UUID, new ApiError(INVALID_REQUEST,
+            "Neither topic name nor id were specified."))
+        } else {
+          maybeAppendToIdsToResolve(topic.topicId)
+        }
+      } else {
+        if (topic.topicId.equals(ZERO_UUID)) {
+          maybeAppendToTopicNamesToResolve(topic.name)
+        } else {
+          appendResponse(topic.name, topic.topicId, new ApiError(INVALID_REQUEST,
+            "You may not specify both topic name and topic id."))
+        }
+      }
+    }
+
+    val idToName = new util.HashMap[Uuid, String]
+    val unknownTopicNameErrors = new util.HashMap[String, ApiError]
+    def maybeAppendToIdToName(id: Uuid, name: String): Unit = {
+      if (duplicatedTopicIds.contains(id) || idToName.put(id, name) != null) {
+          appendResponse(name, id, new ApiError(INVALID_REQUEST,
+              "The same topic was specified by name and by id."))
+          idToName.remove(id)
+          duplicatedTopicIds.add(id)
+      }
+    }
+    controller.findTopicIds(topicNamesToResolve).get().forEach { (name, idOrError) =>
+      if (idOrError.isError)
+        unknownTopicNameErrors.put(name, idOrError.error)
+      else
+        maybeAppendToIdToName(idOrError.result, name)
+    }
+
+    /**
+     * There are 6 error cases to handle here if we don't have permission to delete:
+     *
+     * 1. ID provided, topic missing => UNKNOWN_TOPIC_ID
+     * 2. ID provided, topic present, describeable => TOPIC_AUTHORIZATION_FAILED
+     * 3. ID provided, topic present, undescribeable => UNKNOWN_TOPIC_ID
+     * 4. name provided, topic missing, undescribable => UNKNOWN_TOPIC_OR_PARTITION
```

Comment:
These cases seem wrong. It should be the following:
```
     * 4. name provided, topic missing, undescribable => TOPIC_AUTHORIZATION_FAILED
     * 5. name provided, topic missing, describable => UNKNOWN_TOPIC_OR_PARTITION
     * 6. name provided, topic exists, undescribable => TOPIC_AUTHORIZATION_FAILED
```
If the client does not have describe permission, then it should get `TOPIC_AUTHORIZATION_FAILED` regardless of existence.